pkg/cloudprovider/cloud.go

```diff
@@ -85,13 +85,15 @@ type LoadBalancer interface {
 	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
 	GetLoadBalancer(clusterName string, service *api.Service) (status *api.LoadBalancerStatus, exists bool, err error)
 	// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
-	// Implementations must treat the *api.Service parameter as read-only and not modify it.
+	// Implementations must treat the *api.Service and *api.Node
+	// parameters as read-only and not modify them.
 	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
-	EnsureLoadBalancer(clusterName string, service *api.Service, nodeNames []string) (*api.LoadBalancerStatus, error)
+	EnsureLoadBalancer(clusterName string, service *api.Service, nodes []*api.Node) (*api.LoadBalancerStatus, error)
```

Comment:
I think the concept the cloudprovider API is missing is an "Instance" interface, that would be implemented by each cloudprovider and hold any additional state.  But that brings potential problems in terms of managing the lifespan of those instances.

I'm not sure it make a lot of sense to force the implementation to be more complicated (caching cloud provider ids -> instance data), just so that we can unit test it more easily :-) 

Passing api.Node feels like a great middle ground.

Also, if we ever want to split out the cloudprovider API, having it work in terms of versioned API objects feels like something we know how to deal with.